### PR TITLE
Use lock directory to prevent race conditions

### LINKED ISSUE #5226: Font cache thread safety
The tests for the LSST software stack are run in parallel. We were getting failures roughly 10% of the time because of font cache directory creation issues. Looking at the code in `font_manager.py` we realised that setting `$TRAVIS` would fix our problem, which it did. This feels wrong though. It seems to me that this TRAVIS-specific bandaid is masking a more general problem with the cache. At the very least the environment variable being used to work around the problem should not be TRAVIS as the problem occurs any time two processes are trying to run matplotlib at the same time. Trying to explain why our test harness is forcing `export TRAVIS=1` is challenging :smile:

This is the trace without TRAVIS being set:

```python
   import matplotlib.colorbar
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/colorbar.py", line 34, in <module>
    import matplotlib.collections as collections
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/collections.py", line 27, in <module>
    import matplotlib.backend_bases as backend_bases
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 56, in <module>
    import matplotlib.textpath as textpath
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/textpath.py", line 19, in <module>
    import matplotlib.font_manager as font_manager
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1415, in <module>
    _rebuild()
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1402, in _rebuild
    pickle_dump(fontManager, _fmcache)
  File "/home/vagrant/stack/Linux64/anaconda/2.2.0/lib/python2.7/site-packages/matplotlib/font_manager.py", line 952, in pickle_dump
    with open(filename, 'wb') as fh:
IOError: [Errno 2] No such file or directory: u'/tmp/matplotlib-vagrant/fontList.cache'
```

### PR DESCRIPTION
**Pull request by mdboom**
Creation of the font cache has race conditions.

~~Adds a dependency on lockfile.  We could vendor it if we want to, but the last thing I want to do is maintain cross-platform lockfile code...~~

Fix #5226.

### CODE REVIEW

**Comment by WeatherGod** [comment]
The page for the lockfile package says that the package is deprecated and
recommends using fasteners: http://fasteners.readthedocs.org/en/latest/

**Comment by mdboom** [comment]
Thanks for catching.  What a pain, though.  lockfile has been around forever and is well-tested (and actually has received updates more recently than fasteners) and does what we need and no more.  fasteners has a lot more extra stuff -- but I guess if that's the future, that's where we are.

**Comment by mdboom** [comment]
I don't think fasteners is even the appropriate tool here.  It provides for *thread* locks or *interprocess* locks, but not file system locks for processes that don't even know about each other (which is our use case).  I think it's odd to recommend a package that doesn't even do the same thing.  I think I'd prefer to stick with `filelock` here.

**Comment by WeatherGod** [comment]
Agreed. The alternative package that they recommend is oslo.concurrency
which seems pretty nice (has decorators and such), but I don't know if I
would have to install all of oslo to get just the concurrency part. Might
be a bit too heavy-handed for our needs. We can probably just stick with
lockfile for now, keeping our eyes open for an alternative later?

**Comment by efiring** [comment]
On 2015/10/19 5:57 AM, Michael Droettboom wrote:
> I don't think fasteners is even the appropriate tool here. It provides
> for /thread/ locks or /interprocess/ locks, but not file system locks
> for processes that don't even know about each other (which is our use
> case). I think it's odd to recommend a package that doesn't even do the
> same thing. I think I'd prefer to stick with |filelock| here.

Looking at the documentation, I don't see how the interprocess lock is 
restricted to processes that know about each other.  One can use a 
single decorator that takes a single path argument.

**Comment by mdboom** [comment]
It would seem from Travis-CI that the other problem with lockfile is that it doesn't install cleanly on Python 2.x.  I'll try again with fasteners.

**Comment by pelson** [comment]
> We could vendor it if we want to, but the last thing I want to do is maintain cross-platform lockfile code...

Conda have a tried and tested lockfile implementation that we could lift... https://github.com/conda/conda/blob/master/conda/lock.py - surely this must work on all their supported platforms.

**Comment by mdboom** [comment]
Thanks, Phil.  That might work.  My understanding is that the mkdir approach is the only cross-platform one (there are other, better options on *nix but those don't work on Windows).

File: lib/matplotlib/cbook.py [hunk 1]
```diff
@@ -2550,3 +2550,61 @@ def get_label(y, default_name):
...
+        self.lock_path = os.path.join(self.path, LOCKFN + self.end)
+        self.pattern = os.path.join(self.path, LOCKFN + '-*')
+        self.remove = True
+
+    def __enter__(self):
+        retries = 10
+        # Keep the string "LOCKERROR" in this string so that external
+        # programs can look for it.
+        lockstr = ("""\
+    LOCKERROR: It looks like conda is already doing something.
```
Comment (tacaswell) [inline lib/matplotlib/cbook.py]: Should probably change this to not say conda.

File: lib/matplotlib/font_manager.py [hunk 2]
```diff
@@ -1419,8 +1418,13 @@ def findfont(prop, fontext='ttf'):
 
     def _rebuild():
         global fontManager
+
         fontManager = FontManager()
+
         if _fmcache:
+            with cbook.Locked(_fmcache):
```
Comment (tacaswell) [inline lib/matplotlib/font_manager.py]: should use just `cachedir` here.  It looks like this is creating a folder with the relevant name.

File: lib/matplotlib/font_manager.py [hunk 3]
```diff
@@ -1419,8 +1418,13 @@ def findfont(prop, fontext='ttf'):
...
+                json_dump(fontManager, _fmcache)
```
Comment (tacaswell) [inline lib/matplotlib/font_manager.py]: There is an extra call to `json_dump` in here?

File: lib/matplotlib/cbook.py (hunk 1 above)
Comment (mdboom) [inline lib/matplotlib/cbook.py]: Actually, I'll just take this out -- I took out the message that was displayed here anyway.  I don't think the message is terribly useful in our use case.

File: lib/matplotlib/font_manager.py (hunk 2 above)
Comment (mdboom) [inline lib/matplotlib/font_manager.py]: Yep, that's right.

File: lib/matplotlib/font_manager.py (hunk 3 above)
Comment (mdboom) [inline lib/matplotlib/font_manager.py]: Oops.

**Comment by mdboom** [comment]
Sorry for the quality of this PR.  I was working on this between millions of test runs on another PR.  I keep forgetting I'm not much of a multitasking OS...

**Comment by mdboom** [comment]
Ok -- I've address the comments.  I also squashed the commits, because this went on a number of divergent paths to get here...

**Comment by mdboom** [comment]
I think this is good to merge.  The old hacks to get around this bug have been removed, and Travis-CI is still happy.

**Comment by jenshnielsen** [comment]
Lets see how travis behaves

**Comment by timj** [comment]
Is this patch going to make it in to the 1.5.x release? I don't see it on that branch although I may just be looking in the wrong place.

**Comment by mdboom** [comment]
It wasn't milestoned for that -- there's a fairly high probability for unintended consequences here.  But I could be convinced otherwise if it gets enough testing.

**Comment by timj** [comment]
ok, we can work with the `TRAVIS=1` hack in the interim.

**Comment by jenshnielsen** [comment]
Backported to 2.x as 56a932c4e167a0e7e10ce8457f0f14d1965c0f87

**Pull request merged**